util/cli/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+use clap::{CommandFactory, FromArgMatches, Parser};
+
+/// Command line parser trait that provides build information into the version
+/// string. In order to use this, one must pass in the `version` argument to the
+/// `clap` attribute.  When the `version` flag is present it will be extended to
+/// include the git commit of the crate.
+pub trait ParserWithBuildInfo: Parser {
+    /// Returns the `clap::Command` with the long version appended with the git
+    /// commit information
+    ///
+    /// # Arguments
+    ///
+    /// * `build_info` - A String that will be populated with the build info
+    ///   (git commit). This is passed in as it needs to live as long as the
+    ///   `clap::Command`.
+    fn command_with_build_info(build_info: &mut String) -> clap::Command {
+        let mut command = <Self as CommandFactory>::command();
+        if let Some(version) = command.get_version() {
+            *build_info = format!("{} commit: {}", version, mc_util_build_info::git_commit());
+            command = command.long_version(&**build_info);
+        }
+        command
```

Comment:
This is probably my inexperience with lifetimes.  I did
```
    fn command_with_build_info<'a>() -> (clap::Command<'a>, String){
        let mut command = <Self as CommandFactory>::command();
        let mut build_info = String::new();
        if let Some(version) = command.get_version() {
            build_info = format!("{} commit: {}", version, mc_util_build_info::git_commit());
            command = command.long_version(&*build_info);
        }
        (command, build_info)
    }
```
but that resulted in 
```
$ cargo test --package mc-util-cli
   Compiling mc-util-cli v0.1.0 (/home/nick/git/mobilecoin/util/cli)
error[E0515]: cannot return value referencing local variable `build_info`
  --> util/cli/src/lib.rs:25:9
   |
23 |             command = command.long_version(&*build_info);
   |                                              ---------- `build_info` is borrowed here
24 |         }
25 |         (command, build_info)
   |         ^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function

error[E0505]: cannot move out of `build_info` because it is borrowed
  --> util/cli/src/lib.rs:25:19
   |
18 |     fn command_with_build_info<'a>() -> (clap::Command<'a>, String){
   |                                -- lifetime `'a` defined here
...
23 |             command = command.long_version(&*build_info);
   |                                              ---------- borrow of `build_info` occurs here
24 |         }
25 |         (command, build_info)
   |         ----------^^^^^^^^^^-
   |         |         |
   |         |         move out of `build_info` occurs here
   |         returning this value requires that `build_info` is borrowed for `'a`
   ```

Reply:
Ah, I guess not then